src/serac/numerics/functional/domain_integral.hpp

```diff
@@ -50,59 +52,54 @@ class DomainIntegral {
-        domain_integral::evaluation_kernel<geometry, test_space, trial_space, Q>(U, R, qf_derivatives, J, X,
-                                                                                 num_elements, qf, data);
-      };
-
-      action_of_gradient_ = [qf_derivatives, num_elements, &J](const mfem::Vector& dU, mfem::Vector& dR) {
-        domain_integral::action_of_gradient_kernel<geometry, test_space, trial_space, Q>(dU, dR, qf_derivatives, J,
+      KernelConfig<Q, geometry, test, trials...> eval_config;
+
+      evaluation_ = EvaluationKernel{eval_config, J, X, num_elements, qf, data};
+
+      for_constexpr<num_trial_spaces>([this, num_elements, quadrature_points_per_element, &J, &X, &qf, &data,
+                                       eval_config](auto i) {
+        // allocate memory for the derivatives of the q-function at each quadrature point
+        //
+        // Note: ptrs' lifetime is managed in an unusual way! It is captured by-value in the
+        // action_of_gradient functor below to augment the reference count, and extend its lifetime to match
+        // that of the DomainIntegral that allocated it.
+        using which_trial_space = typename serac::tuple_element<i, serac::tuple<trials...> >::type;
+        using derivative_type   = decltype(get_derivative_type<i, dim, trials...>(qf, data(0, 0)));
+        auto ptr = accelerator::make_shared_array<exec, derivative_type>(num_elements * quadrature_points_per_element);
+        ExecArrayView<derivative_type, 2, exec> qf_derivatives(ptr.get(), num_elements, quadrature_points_per_element);
+
+        evaluation_with_AD_[i] =
+            EvaluationKernel{DerivativeWRT<i>{}, eval_config, qf_derivatives, J, X, num_elements, qf, data};
+
+        // note: this lambda function captures ptr by-value to extend its lifetime
+        //                        vvv
+        action_of_gradient_[i] = [ptr, qf_derivatives, num_elements, J](const mfem::Vector& dU, mfem::Vector& dR) {
+          domain_integral::action_of_gradient_kernel<geometry, test, which_trial_space, Q>(dU, dR, qf_derivatives, J,
+                                                                                           num_elements);
+        };
+
+        element_gradient_[i] = [qf_derivatives, num_elements, J](CPUArrayView<double, 3> K_e) {
+          domain_integral::element_gradient_kernel<geometry, test, which_trial_space, Q>(K_e, qf_derivatives, J,
                                                                                          num_elements);
-      };
-
-      element_gradient_ = [qf_derivatives, num_elements, &J](CPUArrayView<double, 3> K_e) {
-        domain_integral::element_gradient_kernel<geometry, test_space, trial_space, Q>(K_e, qf_derivatives, J,
-                                                                                       num_elements);
-      };
+        };
+      });
     }
 
+#if 0
```

Comment:
This needs an in-code comment stating why it is if-def'd out

Reply:
```suggestion
// some of the GPU functionality is temporarily disabled to 
// help incrementally roll-out the variadic implementation of Functional
// TODO: re-enable GPU kernels in a follow-up PR
#if 0
```